src/cunumeric/stat/histogram_impl.h

```diff
@@ -0,0 +1,140 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+namespace cunumeric {
+namespace detail {
+
+template <typename exe_policy_t, typename array_t, typename bin_t>
+struct lower_bound_op_t {
+  lower_bound_op_t(exe_policy_t, bin_t const* p_bins, size_t n_intervs)
+    : p_bins_(p_bins), n_intervs_(n_intervs)  // CTAD
+  {
+  }
+  __host__ __device__ bool operator()(array_t left, bin_t right) const
+  {
+    // sentinel logic accounts for comparison
+    // against last bin's upper bound, when
+    // (<) is to be replaced by (<=):
+    //
+    auto sentinel = p_bins_[n_intervs_];
+    if constexpr (std::is_same_v<array_t, __half> && (!std::is_integral_v<bin_t>)) {
+      // upcast operands:
+      //
+      bin_t left_up  = static_cast<bin_t>(left);
+      bin_t right_up = static_cast<bin_t>(right);
+      if (left_up == sentinel && right_up == sentinel)
+        return true;
+      else
+        return left_up < right_up;
+    } else if (std::is_same_v<array_t, __half> && std::is_integral_v<bin_t>) {
+      // upcast sentinel:
+      //
+      array_t sentinel_up = static_cast<array_t>(sentinel);
+      if (left == sentinel_up && right == sentinel_up)
+        return true;
+      else
+        return left < right;
+    } else {
+      if (left == sentinel && right == sentinel)
+        return true;
+      else
+        return left < right;
+    }
+  }
+
+ private:
+  bin_t const* p_bins_;
+  size_t n_intervs_;
+};
+
+template <typename exe_policy_t, typename bin_t, typename weight_t>
+struct transform_op_t {
+  transform_op_t(exe_policy_t exe_pol, bin_t const* p_bins, weight_t* p_hist, size_t n_intervs)
+    : ptr_bins_(p_bins)
+  {
+    Sw_ = thrust::reduce(exe_pol, p_hist, p_hist + n_intervs);
+  }
+
+  __host__ __device__ weight_t operator()(size_t index, weight_t h_i)
+  {
+    auto d_i = ptr_bins_[index + 1] - ptr_bins_[index];
+
+#ifdef _ZERO_WIDTH_RET_ZERO_
+    if (d_i == 0)
+      return static_cast<weight_t>(0);
+    else
+#endif
+      return h_i / (Sw_ * d_i);
+  }
+
+ private:
+  bin_t const* ptr_bins_;
+  weight_t* ptr_h_;
+  weight_t Sw_{0};
+};
+
+template <typename exe_policy_t,
+          typename array_t,
+          typename bin_t,
+          template <typename...> typename alloc_t = allocator_t,
```

Comment:
So, this turned out to be necessary, because:
1. Different types are being allocated (e.g., `alloc_t<weight_t, exe_policy_t> alloc_w;` hence parameterized by the type of what is being allocated);
2. For debug purposes I had to occasionally switch to an ad-hoc (say, device) allocator (e.g., using a `thrust::device_vector` underneath, which needs more than one template param to deduce; without the template-template argument deduction fails)